File: introduction/COLSd.m
% Electricity Generators data set from SL79-80
% Corrected OLS for deterministic frontier model

%warning ('off', clear)

%% 1. Import data

data=xlsread('~/Documents/Matlab/DataSets/cowing.xlsx');

global y x1 x2 x3 p1 p2 p3 

y=data(:,2); % output and inputs are presumably in logs
x1=data(:,3);
x2=data(:,4);
x3=data(:,5);
p1 = log(data(:,6)); % original prices are NOT logs
p2 = log(data(:,7));
p3 = log(data(:,8));

%% Estimation 

colsd = LinearModel.fit([x1 x2 x3], y)
resid = colsd.Residuals.Raw;
u_star = - (resid - max(resid));
eff_colsd = exp(-u_star);
hist(eff_colsd);

File: introduction/COLSd.py
# Electricity Generators data set from SL79-80
# Corrected OLS for deterministic frontier model

import numpy                   as np
import pandas                  as pd
import matplotlib.pyplot       as mp
import statsmodels.formula.api as sm

from sklearn import linear_model


#~~~~~~  1. Import data  ~~~~~~#

SOURCE = "../DataSets/cowing.xlsx"
df = pd.read_excel(SOURCE)

for col in ["P1", "P2", "P3"]:
    df[col] = np.log(df[col])
    

#~~~~~~  2. Estimation   ~~~~~~#

pred_model = linear_model.LinearRegression()
pred_model.fit(df[["X1", "X2", "X3"]], df["y"])
pred       = pred_model.predict(df[["X1", "X2", "X3"]])

outp_model = sm.ols(formula = 'y ~ X1 + X2 + X3', data=df[["y", "X1", "X2", "X3"]])
fitted     = outp_model.fit()

resid     = df["y"] - pred
u_star    = -(resid - resid.max())
eff_colsd = np.exp(-u_star)


#~~~~~~  3. Output       ~~~~~~#

print("colsd = \nLinear regression model:\n\ty ~ X1 + X2 + X3")
print("\nEstimated Coefficients:")
print(fitted.summary())

mp.hist(eff_colsd);
